### ISSUE TITLE
Memory corruption in VM

### ISSUE THREAD

**Issue by SkySkimmer** [coq/coq#11170]
Since https://github.com/coq/coq/pull/9914 was closed without merging and has the critical label we should probably have an issue for it ;)

@silene said in https://github.com/coq/coq/pull/9914#issuecomment-553314020 
>I was wrong in my original assessment. Presumably, only the `MAKEACCU` and `ACCUMULATE` opcodes really need to be made safe, as the other ones require the attacker to be malicious. By the way, here is a way to corrupt memory (just increase `1000` to a suitably large value):
> ```coq
> Require Import BinNat.
> Fixpoint T n := match n with O => nat | S n => nat -> T n end.
> Fixpoint app n : T n -> nat :=
>   match n with O => fun x => x | S n => fun f => app n (f O) end.
> Definition n := N.to_nat 1000.
> Axiom f : T n.
> Eval vm_compute in let t := (app n f, 0) in snd t.
> ```
>Anyway, since this is incompatible with #11102, I am closing it.

Critical bug entry from the PR:
~~~
  component: "virtual machine" (compilation to bytecode ran by a C-interpreter)
  summary: buffer overflow on large records and closures
  introduced: V8.1
  impacted released versions: V8.1-V8.9.0
  impacted development branches:
  impacted coqchk versions: none (no virtual machine in coqchk)
  fixed in:
  found by: Dolan, Roux, Melquiond
  exploit:
  GH issue number: ocaml/ocaml#6385
  risk: unlikely to be activated by chance, might happen for autogenerated code
~~~
Thus see also https://github.com/ocaml/ocaml/issues/6385

**Comment by ppedrot** [comment]
This is bad as in "stack smashing and potential arbitrary code execution", not as in "we have a proof of False". (Obviously the former *might* imply the latter, but this probably require more efforts than merely finding another proof of False with template polymorphism.)

In particular, I don't think this should delay the beta release, so postponing.

**Comment by JasonGross** [comment]
What is required for fixing this?  Is it well-understood?  And what's the main bottle-neck?

**Comment by Zimmi48** [comment]
Who would be the right person to address this?

**Comment by Zimmi48** [comment]
Not going to be addressed for 8.12+beta1.

Who did we say we should ping when we discussed this in the Coq Call?

**Comment by Zimmi48** [comment]
Despite my total absence of knowledge on this issue, in my quality of 8.12 RM, I am pinging @proux01, @bgregoir and @xavierleroy since it was said during a ["recent"](https://github.com/coq/coq/wiki/Coq-Call-2020-05-20) Coq Call that they might be able to help.

**Comment by xavierleroy** [comment]
Here is my advice.  First, make sure you've read and understood https://ocaml.org/releases/4.10/htmlman/intfc.html , especially 20.4.4 and 20.5.2.  It says that "alloc small" can only be used with sizes <= `Max_young_wosize`, and `caml_alloc_shr` must be used for bigger blocks.

To handle an arbitrary size, you therefore need to distinguish on the size, like here for example:
https://github.com/ocaml/ocaml/blob/bd510cdf93628952750fbf52ef5ee8cf05c4e721/runtime/interp.c#L674-L689

If you don't want to handle arbitrary sizes, e.g. you claim that MAKEBLOCK can use `Alloc_small` unconditionally because nobody in their right minds should have a constructor with more than `Max_young_wosize` arguments, it's your choice, but then by all means add a compile-time error in addition to a run-time assertion in the VM.  The compile-time error will be much easier to understand by the user.

Hope this helps.





Second,

**Comment by bgregoir** [comment]
> Le 23 juin 2020 à 17:48, Xavier Leroy <notifications@github.com> a écrit :
> 
> 
> Here is my advice. First, make sure you've read and understood https://ocaml.org/releases/4.10/htmlman/intfc.html <https://ocaml.org/releases/4.10/htmlman/intfc.html> , especially 20.4.4 and 20.5.2. It says that "alloc small" can only be used with sizes <= Max_young_wosize, and caml_alloc_shr must be used for bigger blocks.
> 
> To handle an arbitrary size, you therefore need to distinguish on the size, like here for example:
> https://github.com/ocaml/ocaml/blob/bd510cdf93628952750fbf52ef5ee8cf05c4e721/runtime/interp.c#L674-L689 <https://github.com/ocaml/ocaml/blob/bd510cdf93628952750fbf52ef5ee8cf05c4e721/runtime/interp.c#L674-L689>
> If you don't want to handle arbitrary sizes, e.g. you claim that MAKEBLOCK can use Alloc_small unconditionally because nobody in their right minds should have a constructor with more than Max_young_wosize arguments, it's your choice, but then by all means add a compile-time error in addition to a run-time assertion in the VM. The compile-time error will be much easier to understand by the user.
> 
> Hope this helps.
> 
> 
Hi Xavier,
   I think here the problem cannot be detected at compile time. 
   The function build an « accumulate » block with 1000 arguments so I think we should do modify the ACCUMULATE instruction
    such that it will use caml_alloc_shr when the number of argument become to big.
    For constructor with more that Max_young_wosize, we can do it at compile time.
    I will try to fix that.
B

**Comment by proux01** [comment]
I agree the simplest solution seems to be adapting `coq_interp.c` following Xavier's example. As far as I see, the following instructions need a fix: `GRAB`, `GRABREC`, `CLOSURE`, `CLOSUREREC`, `CLOSURECOFIX`, `MAKEBLOCK`, `ACCUMULATE`, `MAKESWITCHBLOCK` and `MAKEACCU`. Note that there is also a strong assumption in function `coq_tcode_array` in `coq_values.c`.

**Comment by gares** [comment]
Postponing, postponing... this playing the ball game is not helping managing releases.
I'll ping one last time @silene @proux01 @ppedrot @SkySkimmer and @maximedenes , IMO you guys are the ones that can fix this problem (I can't, sorry). If you don't have time in the (very) near future, then I'll just clear the milestone.

**Comment by ppedrot** [comment]
It seems @bgregoir was going to work on it, so pinging him again is probably a good strategy.

**Comment by SkySkimmer** [comment]
>IMO you guys are the ones that can fix this problem

Not me imo

**Comment by silene** [comment]
To correct what is quoted in the first post, `GRABREC` (irreducible case) also needs to be fixed:
```ocaml
num_args = coq_extra_args - rec_pos;
Alloc_small(block, 3 + num_args, Closure_tag);
```

And as already mentioned, there is `ACCUMULATE`:
```ocaml
size = Wosize_val(coq_env);
Alloc_small(accu, size + coq_extra_args + 1, Closure_tag);
```

and `MAKEACCU`:
```ocaml
Alloc_small(accu, coq_extra_args + 4, Closure_tag);
```

The three above have to be handled at runtime, because they can happen at any time.

And then there are all of those that I put in the "malicious user" category. They should either be forbidden at compile time or handled at runtime: `GRAB`, `GRABREC` (closure case), `CLOSURE`, `CLOSUREREC`, `CLOSURECOFIX`, `MAKEBLOCK`, `MAKESWITCHBLOCK`. Contrarily to the above three, they can only be caused by the user writing inductive types with huge numbers of constructors, constructors with huge numbers of arguments, functions with huge numbers of arguments, and recursive functions with huge numbers of siblings. (To be clear, as shown in the OCaml case, this can actually happen even if the user is not malicious, in case of automatically generated code.)

I can take care of the three unfortunate ones. But I don't feel like tackling all the malicious ones right now.

**Comment by xavierleroy** [comment]
> I can take care of the three unfortunate ones.

I can review a PR along these lines.  Just @ - mention me.

**Comment by SkySkimmer** [comment]
The PR which closed this said "partially fix" so this should probably still be open

**Comment by silene** [comment]
Note that #13439 covers all the issues not fixed by #13431.
